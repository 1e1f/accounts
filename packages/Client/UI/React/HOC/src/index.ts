export { default as Provider } from './Provider';
export { default as withAccounts } from './withAccounts';