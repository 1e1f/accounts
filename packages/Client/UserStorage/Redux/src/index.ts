export { default } from './UserStorageRedux';
export { default as reducer } from './reducer'