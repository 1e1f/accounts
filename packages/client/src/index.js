// @flow

import Accounts, { AccountsClient } from './AccountsClient';
import type { TransportInterface } from './TransportInterface';
import config from './config';
import reducer from './module';

export default Accounts;

export {
  AccountsClient,
  config,
  reducer,
};

export type {
  TransportInterface,
};
