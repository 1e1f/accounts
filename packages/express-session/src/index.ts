import { AccountsServer } from '@accounts/server';
import { Tokens, User } from '@accounts/types';
import { Request, Response, NextFunction } from 'express';
import * as requestIp from 'request-ip';
import 'express-session';
import { getUserAgent } from './utils/get-user-agent';

export interface AccountsSessionOptions {
  user: {
    name: string;
    resolve: (tokens: Tokens) => User | Promise<User>;
  };
  name?: string;
}

export class AccountsSession {
  constructor(private accountsServer: AccountsServer, private options: AccountsSessionOptions) {
    if (!this.options.name) {
      this.options.name = 'js-accounts-tokens';
    }
  }

  public middleware() {
    return async (req: Request, res: Response, next: NextFunction) => {
      try {
        await this.renew(req);

        const tokens = this.get(req);
        const user = await this.options.user.resolve(tokens);

        req[this.options.user.name] = user;

        next();
      } catch (e) {
        next(e);
      }
    };
  }

  public async destroy(req?: Request, callback?: (err: any) => void): Promise<void> {
    req.session.destroy(async err => {
      const tokens = this.get(req);

      if (tokens && tokens.accessToken) {
        await this.accountsServer.logout(tokens.accessToken);
        await this.clear(req);
      }

      if (callback) {
        await callback(err);
      }
    });
  }

  public get(req?: Request): Tokens | undefined {
    if (!req) {
      return;
    }

    if (req.session && req.session[this.options.name]) {
      return req.session[this.options.name];
    }
  }

  public async renew(req: Request): Promise<Tokens | undefined> {
    const tokens = this.get(req);

    if (this.accountsServer && tokens && tokens.accessToken && tokens.refreshToken) {
      const result = await this.accountsServer.refreshTokens(
        tokens.accessToken,
        tokens.refreshToken,
        requestIp.getClientIp(req),
        getUserAgent(req)
      );

      this.set(req, result.tokens);

      return result.tokens;
    }
  }

  public set(req: Request, tokens: Tokens): void {
    if (!tokens) {
      this.clear(req);
      return;
    }

    if (req.session) {
      req.session[this.options.name] = tokens;
    }
  }

  public clear(req: Request): void {
    if (this.get(req)) {
      req.session[this.options.name] = null;
    }
  }
}
