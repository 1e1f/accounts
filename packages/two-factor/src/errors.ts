export const errors = {
  codeDidNotMatch: `2FA code didn't match`,
};
