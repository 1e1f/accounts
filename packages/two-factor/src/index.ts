export { TwoFactor } from './two-factor';