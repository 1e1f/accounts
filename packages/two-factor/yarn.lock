# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
# yarn lockfile v1


"@accounts/common@^0.1.0-beta.3":
  version "0.1.0-beta.3"
  resolved "https://registry.yarnpkg.com/@accounts/common/-/common-0.1.0-beta.3.tgz#311a9676526160b37fcb8aec3526f3d0e0cad9e6"
  dependencies:
    lodash "^4.16.4"

"@accounts/server@^0.1.0-beta.3":
  version "0.1.0-beta.3"
  resolved "https://registry.yarnpkg.com/@accounts/server/-/server-0.1.0-beta.3.tgz#fc584094bedce26c115d9e19679dcf0259a4a675"
  dependencies:
    "@accounts/common" "^0.1.0-beta.3"
    babel-polyfill "^6.23.0"
    bcryptjs "^2.4.0"
    jsonwebtoken "^8.0.0"
    jwt-decode "^2.1.0"
    lodash "^4.16.4"

"@types/speakeasy@^2.0.2":
  version "2.0.2"
  resolved "https://registry.yarnpkg.com/@types/speakeasy/-/speakeasy-2.0.2.tgz#153ec3636eea0562209b0a2f1fdf8b479286919b"

babel-polyfill@^6.23.0:
  version "6.26.0"
  resolved "https://registry.yarnpkg.com/babel-polyfill/-/babel-polyfill-6.26.0.tgz#379937abc67d7895970adc621f284cd966cf2153"
  dependencies:
    babel-runtime "^6.26.0"
    core-js "^2.5.0"
    regenerator-runtime "^0.10.5"

babel-runtime@^6.26.0:
  version "6.26.0"
  resolved "https://registry.yarnpkg.com/babel-runtime/-/babel-runtime-6.26.0.tgz#965c7058668e82b55d7bfe04ff2337bc8b5647fe"
  dependencies:
    core-js "^2.4.0"
    regenerator-runtime "^0.11.0"

base32.js@0.0.1:
  version "0.0.1"
  resolved "https://registry.yarnpkg.com/base32.js/-/base32.js-0.0.1.tgz#d045736a57b1f6c139f0c7df42518a84e91bb2ba"

base64url@2.0.0, base64url@^2.0.0:
  version "2.0.0"
  resolved "https://registry.yarnpkg.com/base64url/-/base64url-2.0.0.tgz#eac16e03ea1438eff9423d69baa36262ed1f70bb"

bcryptjs@^2.4.0:
  version "2.4.3"
  resolved "https://registry.yarnpkg.com/bcryptjs/-/bcryptjs-2.4.3.tgz#9ab5627b93e60621ff7cdac5da9733027df1d0cb"

buffer-equal-constant-time@1.0.1:
  version "1.0.1"
  resolved "https://registry.yarnpkg.com/buffer-equal-constant-time/-/buffer-equal-constant-time-1.0.1.tgz#f8e71132f7ffe6e01a5c9697a4c6f3e48d5cc819"

core-js@^2.4.0, core-js@^2.5.0:
  version "2.5.3"
  resolved "https://registry.yarnpkg.com/core-js/-/core-js-2.5.3.tgz#8acc38345824f16d8365b7c9b4259168e8ed603e"

ecdsa-sig-formatter@1.0.9:
  version "1.0.9"
  resolved "https://registry.yarnpkg.com/ecdsa-sig-formatter/-/ecdsa-sig-formatter-1.0.9.tgz#4bc926274ec3b5abb5016e7e1d60921ac262b2a1"
  dependencies:
    base64url "^2.0.0"
    safe-buffer "^5.0.1"

jsonwebtoken@^8.0.0:
  version "8.2.0"
  resolved "https://registry.yarnpkg.com/jsonwebtoken/-/jsonwebtoken-8.2.0.tgz#690ec3a9e7e95e2884347ce3e9eb9d389aa598b3"
  dependencies:
    jws "^3.1.4"
    lodash.includes "^4.3.0"
    lodash.isboolean "^3.0.3"
    lodash.isinteger "^4.0.4"
    lodash.isnumber "^3.0.3"
    lodash.isplainobject "^4.0.6"
    lodash.isstring "^4.0.1"
    lodash.once "^4.0.0"
    ms "^2.1.1"
    xtend "^4.0.1"

jwa@^1.1.4:
  version "1.1.5"
  resolved "https://registry.yarnpkg.com/jwa/-/jwa-1.1.5.tgz#a0552ce0220742cd52e153774a32905c30e756e5"
  dependencies:
    base64url "2.0.0"
    buffer-equal-constant-time "1.0.1"
    ecdsa-sig-formatter "1.0.9"
    safe-buffer "^5.0.1"

jws@^3.1.4:
  version "3.1.4"
  resolved "https://registry.yarnpkg.com/jws/-/jws-3.1.4.tgz#f9e8b9338e8a847277d6444b1464f61880e050a2"
  dependencies:
    base64url "^2.0.0"
    jwa "^1.1.4"
    safe-buffer "^5.0.1"

jwt-decode@^2.1.0:
  version "2.2.0"
  resolved "https://registry.yarnpkg.com/jwt-decode/-/jwt-decode-2.2.0.tgz#7d86bd56679f58ce6a84704a657dd392bba81a79"

lodash.includes@^4.3.0:
  version "4.3.0"
  resolved "https://registry.yarnpkg.com/lodash.includes/-/lodash.includes-4.3.0.tgz#60bb98a87cb923c68ca1e51325483314849f553f"

lodash.isboolean@^3.0.3:
  version "3.0.3"
  resolved "https://registry.yarnpkg.com/lodash.isboolean/-/lodash.isboolean-3.0.3.tgz#6c2e171db2a257cd96802fd43b01b20d5f5870f6"

lodash.isinteger@^4.0.4:
  version "4.0.4"
  resolved "https://registry.yarnpkg.com/lodash.isinteger/-/lodash.isinteger-4.0.4.tgz#619c0af3d03f8b04c31f5882840b77b11cd68343"

lodash.isnumber@^3.0.3:
  version "3.0.3"
  resolved "https://registry.yarnpkg.com/lodash.isnumber/-/lodash.isnumber-3.0.3.tgz#3ce76810c5928d03352301ac287317f11c0b1ffc"

lodash.isplainobject@^4.0.6:
  version "4.0.6"
  resolved "https://registry.yarnpkg.com/lodash.isplainobject/-/lodash.isplainobject-4.0.6.tgz#7c526a52d89b45c45cc690b88163be0497f550cb"

lodash.isstring@^4.0.1:
  version "4.0.1"
  resolved "https://registry.yarnpkg.com/lodash.isstring/-/lodash.isstring-4.0.1.tgz#d527dfb5456eca7cc9bb95d5daeaf88ba54a5451"

lodash.once@^4.0.0:
  version "4.1.1"
  resolved "https://registry.yarnpkg.com/lodash.once/-/lodash.once-4.1.1.tgz#0dd3971213c7c56df880977d504c88fb471a97ac"

lodash@^4.16.4:
  version "4.17.5"
  resolved "https://registry.yarnpkg.com/lodash/-/lodash-4.17.5.tgz#99a92d65c0272debe8c96b6057bc8fbfa3bed511"

ms@^2.1.1:
  version "2.1.1"
  resolved "https://registry.yarnpkg.com/ms/-/ms-2.1.1.tgz#30a5864eb3ebb0a66f2ebe6d727af06a09d86e0a"

regenerator-runtime@^0.10.5:
  version "0.10.5"
  resolved "https://registry.yarnpkg.com/regenerator-runtime/-/regenerator-runtime-0.10.5.tgz#336c3efc1220adcedda2c9fab67b5a7955a33658"

regenerator-runtime@^0.11.0:
  version "0.11.1"
  resolved "https://registry.yarnpkg.com/regenerator-runtime/-/regenerator-runtime-0.11.1.tgz#be05ad7f9bf7d22e056f9726cee5017fbf19e2e9"

safe-buffer@^5.0.1:
  version "5.1.1"
  resolved "https://registry.yarnpkg.com/safe-buffer/-/safe-buffer-5.1.1.tgz#893312af69b2123def71f57889001671eeb2c853"

speakeasy@^2.0.0:
  version "2.0.0"
  resolved "https://registry.yarnpkg.com/speakeasy/-/speakeasy-2.0.0.tgz#85c91a071b09a5cb8642590d983566165f57613a"
  dependencies:
    base32.js "0.0.1"

xtend@^4.0.1:
  version "4.0.1"
  resolved "https://registry.yarnpkg.com/xtend/-/xtend-4.0.1.tgz#a5c6d532be656e23db820efb943a1f04998d63af"
