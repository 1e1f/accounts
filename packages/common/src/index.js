// @flow
/* eslint-disable import/no-named-as-default */
import * as PasswordSignupFields from './passwordSignupFields';
import * as validators from './validators';
import { AccountsError } from './errors';
import toUsernameAndEmail from './toUsernameAndEmail';
import config from './config';

import type {
  UserObjectType,
  CreateUserType,
  PasswordLoginUserType,
  PasswordLoginUserIdentityType,
  LoginReturnType,
  TokensType,
  SessionType,
} from './types';

export type {
  UserObjectType,
  CreateUserType,
  PasswordLoginUserIdentityType,
  PasswordLoginUserType,
  LoginReturnType,
  TokensType,
  SessionType,
};

export {
  PasswordSignupFields,
  validators,
  AccountsError,
  toUsernameAndEmail,
  config,
};
