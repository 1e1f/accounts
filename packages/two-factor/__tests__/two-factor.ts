import * as speakeasy from 'speakeasy';
import { TwoFactor } from '../src/two-factor';

const dbMock: any = {
  findUserById: jest.fn(),
  setService: jest.fn(),
  unsetService: jest.fn(),
};

const mockedUserWithoutTwoFactor = {
  services: {}
}

const mockedUserWithTwoFactor = {
  services: {
    'two-factor': { secret: {} }
  }
}

const accountsTwoFactor = new TwoFactor({
  appName: 'Accounts-js test',
});
accountsTwoFactor.setStore(dbMock);

describe('TwoFactor', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    jest.restoreAllMocks();
  });

  describe('getNewAuthSecret', () => {
    it('should return a new secret', () => {
      const secret = accountsTwoFactor.getNewAuthSecret();
      expect(secret).toBeTruthy();
    });
  });

  describe('set', () => {
    it('should throw if invalid code', async () => {
      const secret = accountsTwoFactor.getNewAuthSecret();
      await expect(
        accountsTwoFactor.set('userId', secret, 'invalidCode')
      ).rejects.toThrowErrorMatchingSnapshot();
    });

    it('should set two-factor service', async () => {
      const spy = jest.spyOn(speakeasy.totp, 'verify').mockReturnValue(true);
      await accountsTwoFactor.set('userId', 'secret' as any, 'validCode');
      expect(spy).toHaveBeenCalled();
      expect(dbMock.setService.mock.calls).toMatchSnapshot();
    });
  });

  describe('unset', () => {
    it('should throw that 2fa is not set', async () => {
      dbMock.findUserById.mockImplementation(() => Promise.resolve(mockedUserWithoutTwoFactor));
      await expect(
        accountsTwoFactor.unset('userId', 'invalidCode')
      ).rejects.toThrowErrorMatchingSnapshot();
    });

    it('should throw if invalid code', async () => {
      dbMock.findUserById.mockImplementation(() => Promise.resolve(mockedUserWithTwoFactor));
      await expect(
        accountsTwoFactor.unset('userId', 'invalidCode')
      ).rejects.toThrowErrorMatchingSnapshot();
    });

    it('should unset two-factor service', async () => {
      dbMock.findUserById.mockImplementation(() => Promise.resolve(mockedUserWithTwoFactor));
      const spy = jest.spyOn(speakeasy.totp, 'verify').mockReturnValue(true);
      await accountsTwoFactor.unset('userId', 'validCode');
      expect(spy).toHaveBeenCalled();
      expect(dbMock.unsetService.mock.calls).toMatchSnapshot();
      spy.mockRestore();
    });
  });
});
