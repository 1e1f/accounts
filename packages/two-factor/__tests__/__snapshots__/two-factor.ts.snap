// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TwoFactor set should set two-factor service 1`] = `
Array [
  Array [
    "userId",
    "two-factor",
    Object {
      "secret": "secret",
    },
  ],
]
`;

exports[`TwoFactor set should throw if invalid code 1`] = `"2FA code didn't match"`;

exports[`TwoFactor unset should throw if invalid code 1`] = `"2FA code didn't match"`;

exports[`TwoFactor unset should throw that 2fa is not set 1`] = `"2FA not set"`;

exports[`TwoFactor unset should unset two-factor service 1`] = `
Array [
  Array [
    "userId",
    "two-factor",
  ],
]
`;
